extras/kms/examples/cli/main.go

```diff
@@ -0,0 +1,181 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+
+	"github.com/hashicorp/go-dbw"
+	"github.com/hashicorp/go-kms-wrapping/extras/kms/examples/v2"
+	"github.com/hashicorp/go-kms-wrapping/extras/kms/v2"
+	wrapping "github.com/hashicorp/go-kms-wrapping/v2"
+)
+
+const globalScope = "global"
+
+// default to an aead root kms
+const rootKmsAead = `
+kms "aead" {
+	purpose = "root"
+	aead_type = "aes-gcm"
+	key = "%s"
+	key_id = "global_root"
+}`
+
+// support the --use-transit flag which requires the caller to run:
+// "docker-compose up" before executing the example
+const rootKmsTransit = `
+kms "transit" {
+	purpose            = "root"
+	address            = "http://localhost:8200"
+	token              = "vault-plaintext-root-token"
+	disable_renewal    = "false"  
+	key_name           = "examplekey"
+	mount_path         = "transit/"
+	namespace          = "ns1/"
+  }`
+
+func main() {
+	mainCtx := context.Background()
+
+	debug := flag.Bool("debug", false, "enable debug")
+	pt := flag.String("plaintext", "default plaintext secret", "plaintext you'd like to use for encrypt/decrypt ops with a wrapper")
+	useTransit := flag.Bool("use-transit", false, `use vault transit as the root wrapper source - run "docker-compose up" first`)
+	flag.Parse()
+
+	var kmsHcl string
+	switch {
+	case *useTransit:
+		kmsHcl = rootKmsTransit
+	default:
+		key := examples.GenerateKey()
+		kmsHcl = fmt.Sprintf(rootKmsAead, key)
+	}
+
+	// get the root wrapper from the provided configuration hcl
+	rootWrapper, err := examples.RootWrapperFromConfig(mainCtx, kmsHcl, *useTransit)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "unable to get root wrapper from config: %s\n\n", err)
+		return
+	}
+	// open the db and run migrations for both for the kms and this cli app.
+	// the cli app migration adds a scope table and a fk to the kms_root_key
+	// table.
+	rw, err := examples.OpenDB(mainCtx)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to open db: %s\n\n", err)
+		return
+	}
+	if *debug {
+		rw.DB().Debug(true)
+	}
+	// create a kms that supports both the default kms.KeyPurposeRootKey KEK and
+	// a "database" DEK
+	k, err := kms.New(rw, rw, []kms.KeyPurpose{"database"})
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to init kms: %s\n\n", err)
+		return
+	}
+
+	// add the external root key wrapper.
+	if err := k.AddExternalWrapper(mainCtx, kms.KeyPurposeRootKey, rootWrapper); err != nil {
+		fmt.Fprintf(os.Stderr, "failed to add root key: %s\n\n", err)
+		return
+	}
+
+	// Create the global scope and it's related kms wrappers
+	if _, err := rw.DoTx(mainCtx, func(error) bool { return false }, 10, dbw.ExpBackoff{},
+		func(r dbw.Reader, w dbw.Writer) error {
+			if err := w.Create(mainCtx, &examples.Scope{PrivateId: globalScope}); err != nil {
+				return err
+			}
+			if err := k.CreateKeys(mainCtx, globalScope, []kms.KeyPurpose{"database"}, kms.WithReaderWriter(r, w)); err != nil {
+				return err
+			}
+			return nil
```

Comment:
yeah... I've always struggled with this since IMO it's more difficult to read.  The symmetry between the calls makes it easy for me to scan quickly and realize that both Create(...) stmts are handled the same way.  When I see a direct return for the second call my mind just stops to see if there's any real diff between the two.